File: app/src/main/java/com/qimian233/ztool/hook/modules/systemui/BrightnessSliderPercentageHook.java
package com.qimian233.ztool.hook.modules.systemui;

import android.content.Context;
import android.graphics.Color;
import android.graphics.Typeface;
import android.view.View;
import android.view.ViewGroup;
import android.view.ViewParent;
import android.widget.FrameLayout;
import android.widget.ProgressBar;
import android.widget.SeekBar;
import android.widget.TextView;

import com.qimian233.ztool.hook.base.BaseHookModule;
import com.qimian233.ztool.hook.base.PreferenceHelper;

import java.util.Locale;
import java.util.Map;
import java.util.WeakHashMap;

import de.robv.android.xposed.XC_MethodHook;
import de.robv.android.xposed.XposedHelpers;
import de.robv.android.xposed.callbacks.XC_LoadPackage;

public class BrightnessSliderPercentageHook extends BaseHookModule {
    private static final String SYSTEM_UI_PACKAGE = "com.android.systemui";
    private static final String TOGGLE_SLIDER_VIEW_CLASS = "com.android.systemui.settings.ToggleSliderView";
    private static final String SLIDER_PERCENT_TAG = "ztool_control_center_slider_percent";
    private static final String BRIGHTNESS_ROOT_FIELD = "mBrightnessSliderRoot";
    private static final String BRIGHTNESS_ICON_FIELD = "mBrightnessIconMark";
    private static final int LABEL_GAP_DP = 2;
    private static final String PREF_KEY = "brightness_slider_percentage";

    private static boolean PERCENTAGE_ENABLED = false;

    private final Map<View, View.OnLayoutChangeListener> layoutListeners = new WeakHashMap<>();

    @Override
    public String getModuleName() {
        return PREF_KEY;
    }

    @Override
    public String[] getTargetPackages() {
        return new String[]{SYSTEM_UI_PACKAGE};
    }

    @Override
    public void handleLoadPackage(XC_LoadPackage.LoadPackageParam lpparam) throws Throwable {
        updatePrefs();
        hookToggleSliderViewLifecycle(lpparam);
        hookBrightnessControllerCallbacks(lpparam);
        hookSeekProgressChanges();
        log("Brightness slider percentage hooks installed");
    }

    private void hookBrightnessControllerCallbacks(XC_LoadPackage.LoadPackageParam lpparam) {
        XposedHelpers.findAndHookMethod(
                "com.android.systemui.settings.brightness.BrightnessController",
                lpparam.classLoader,
                "onChanged",
                int.class,
                boolean.class,
                boolean.class,
                new XC_MethodHook() {
                    @Override
                    protected void afterHookedMethod(MethodHookParam param) {
                        refreshBrightnessFromController(param.thisObject, (Integer) param.args[0]);
                    }
                }
        );

        XposedHelpers.findAndHookMethod(
                "com.android.systemui.settings.brightness.BrightnessSliderController",
                lpparam.classLoader,
                "setValue",
                int.class,
                new XC_MethodHook() {
                    @Override
                    protected void afterHookedMethod(MethodHookParam param) {
                        Object sliderController = param.thisObject;
                        Object view = XposedHelpers.getObjectField(sliderController, "mView");
                        if (view instanceof View) {
                            refreshBrightnessFromView((View) view, (Integer) param.args[0]);
                        }
                    }
                }
        );
    }

    private void hookToggleSliderViewLifecycle(XC_LoadPackage.LoadPackageParam lpparam) {
        XposedHelpers.findAndHookConstructor(
                TOGGLE_SLIDER_VIEW_CLASS,
                lpparam.classLoader,
                Context.class,
                android.util.AttributeSet.class,
                int.class,
                new XC_MethodHook() {
                    @Override
                    protected void afterHookedMethod(MethodHookParam param) {
                        attachSliderLabel(param.thisObject);
                    }
                }
        );

        XposedHelpers.findAndHookMethod(
                TOGGLE_SLIDER_VIEW_CLASS,
                lpparam.classLoader,
                "updateBrightnessSlider",
                new XC_MethodHook() {
                    @Override
                    protected void afterHookedMethod(MethodHookParam param) {
                        attachSliderLabel(param.thisObject);
                        refreshBrightnessLabel(param.thisObject);
                    }
                }
        );

        XposedHelpers.findAndHookMethod(
                TOGGLE_SLIDER_VIEW_CLASS,
                lpparam.classLoader,
                "refreshSeekBar",
                ProgressBar.class,
                new XC_MethodHook() {
                    @Override
                    protected void afterHookedMethod(MethodHookParam param) {
                        Object sliderView = param.thisObject;
                        ProgressBar progressBar = (ProgressBar) param.args[0];
                        if (isBrightnessSlider(sliderView, progressBar)) {
                            refreshBrightnessLabel(sliderView);
                        }
                    }
                }
        );
    }

    private void hookSeekProgressChanges() {
        XposedHelpers.findAndHookMethod(
                SeekBar.class,
                "setProgress",
                int.class,
                new XC_MethodHook() {
                    @Override
                    protected void afterHookedMethod(MethodHookParam param) {
                        SeekBar seekBar = (SeekBar) param.thisObject;
                        Object sliderView = findToggleSliderView(seekBar);
                        if (sliderView == null) {
                            return;
                        }
                        if (isBrightnessSlider(sliderView, seekBar)) {
                            refreshBrightnessLabel(sliderView);
                        }
                    }
                }
        );
    }

    private void attachSliderLabel(Object sliderView) {
        try {
            attachLabelToRoot(sliderView);
        } catch (Throwable t) {
            logError("Failed to attach brightness slider label", t);
        }
    }

    private void attachLabelToRoot(Object sliderView) {
        FrameLayout root = getFrameLayoutField(sliderView);
        View icon = getViewField(sliderView);
        if (root == null || icon == null) {
            return;
        }

        if (!PERCENTAGE_ENABLED) {
            removePercentView(root);
            return;
        }

        TextView percentView = findPercentView(root);
        if (percentView == null) {
            percentView = createPercentView(root.getContext());
            root.addView(percentView, createLayoutParams());
        }

        ensureLayoutTracking(root, icon, percentView);
        updateBrightnessPercentColor(sliderView, percentView);
        refreshBrightnessLabel(sliderView, percentView);
    }

    private void refreshBrightnessLabel(Object sliderView) {
        if (!PERCENTAGE_ENABLED) {
            detachBrightnessLabel(sliderView);
            return;
        }
        FrameLayout root = getFrameLayoutField(sliderView);
        if (root == null) {
            return;
        }
        TextView percentView = findPercentView(root);
        if (percentView != null) {
            refreshBrightnessLabel(sliderView, percentView);
        }
    }

    private void refreshBrightnessLabel(Object sliderView, TextView percentView) {
        if (!PERCENTAGE_ENABLED) {
            removePercentView(getFrameLayoutField(sliderView));
            return;
        }
        try {
            SeekBar brightnessSlider = (SeekBar) XposedHelpers.getObjectField(sliderView, "mBrightnessSlider");
            if (brightnessSlider == null) {
                percentView.setText("--%");
                return;
            }
            percentView.setText(formatPercent(
                    brightnessSlider.getProgress(),
                    brightnessSlider.getMin(),
                    brightnessSlider.getMax()
            ));
            updateBrightnessPercentColor(sliderView, percentView);
            schedulePositionUpdate(sliderView);
        } catch (Throwable t) {
            percentView.setText("--%");
            if (DEBUG) {
                logError("Failed to refresh brightness percent", t);
            }
        }
    }

    private void refreshBrightnessFromController(Object brightnessController, int progress) {
        if (!PERCENTAGE_ENABLED || brightnessController == null) {
            return;
        }
        try {
            Object control = XposedHelpers.getObjectField(brightnessController, "mControl");
            if (control == null) {
                return;
            }
            Object view = XposedHelpers.getObjectField(control, "mView");
            if (!(view instanceof View)) {
                return;
            }
            refreshBrightnessFromView((View) view, progress);
        } catch (Throwable t) {
            if (DEBUG) {
                logError("Failed to refresh brightness from controller", t);
            }
        }
    }

    private void refreshBrightnessFromView(View view, int progress) {
        if (view == null) {
            return;
        }
        Object sliderView = findToggleSliderView(view);
        if (sliderView == null) {
            return;
        }
        FrameLayout root = getFrameLayoutField(sliderView);
        if (root == null) {
            return;
        }
        TextView percentView = findPercentView(root);
        if (percentView == null) {
            return;
        }

        try {
            SeekBar brightnessSlider = (SeekBar) XposedHelpers.getObjectField(sliderView, "mBrightnessSlider");
            int max = brightnessSlider != null ? brightnessSlider.getMax() : 65535;
            max = Math.max(1, max);
            int percent = Math.max(0, Math.min(100, Math.round((progress * 100f) / max)));
            percentView.setText(String.format(Locale.US, "%d%%", percent));
            updateBrightnessPercentColor(sliderView, percentView);
            schedulePositionUpdate(sliderView);
        } catch (Throwable t) {
            if (DEBUG) {
                logError("Failed to refresh brightness from view", t);
            }
        }
    }

    private void updateBrightnessPercentColor(Object sliderView, TextView percentView) {
        percentView.setTextColor(resolveBrightnessPercentColor(sliderView));
    }

    private int resolveBrightnessPercentColor(Object sliderView) {
        SeekBar brightnessSlider = sliderView instanceof SeekBar
                ? (SeekBar) sliderView
                : (SeekBar) XposedHelpers.getObjectField(sliderView, "mBrightnessSlider");
        if (brightnessSlider == null) {
            return Color.argb(0xff, 0xd8, 0xd8, 0xd8);
        }
        float progress = ((brightnessSlider.getProgress() - brightnessSlider.getMin()) * 1.0f)
                / Math.max(1, brightnessSlider.getMax() - brightnessSlider.getMin());
        if (progress < 0.2f) {
            return Color.argb(0xff, 0xd8, 0xd8, 0xd8);
        }
        int gray = (int) ((1.0f - Math.min((progress - 0.2f) / 0.2f, 1.0f)) * 216.0f);
        gray = Math.max(gray, 0x80);
        int alpha = Math.min(((int) Math.floor(progress * 85.0f)) + 170, 255);
        return Color.argb(alpha, gray, gray, gray);
    }

    private String formatPercent(int progress, int min, int max) {
        int range = Math.max(1, max - min);
        int value = Math.round(((progress - min) * 100f) / range);
        value = Math.max(0, Math.min(100, value));
        return String.format(Locale.US, "%d%%", value);
    }

    private TextView createPercentView(Context context) {
        TextView textView = new TextView(context);
        textView.setTag(SLIDER_PERCENT_TAG);
        textView.setTextColor(Color.argb(0xff, 0xd8, 0xd8, 0xd8));
        textView.setTypeface(Typeface.DEFAULT_BOLD);
        textView.setTextSize(13f);
        textView.setShadowLayer(2f, 0f, 0f, Color.BLACK);
        textView.setSingleLine(true);
        textView.setIncludeFontPadding(false);
        textView.setGravity(android.view.Gravity.CENTER);
        textView.setImportantForAccessibility(View.IMPORTANT_FOR_ACCESSIBILITY_NO);
        textView.setClickable(false);
        textView.setFocusable(false);
        textView.setPadding(dp(context, 6), dp(context, 2), dp(context, 6), dp(context, 2));
        textView.setElevation(dp(context, 2));
        return textView;
    }

    private FrameLayout.LayoutParams createLayoutParams() {
        FrameLayout.LayoutParams params = new FrameLayout.LayoutParams(
                ViewGroup.LayoutParams.WRAP_CONTENT,
                ViewGroup.LayoutParams.WRAP_CONTENT
        );
        params.gravity = android.view.Gravity.TOP | android.view.Gravity.START;
        return params;
    }

    private TextView findPercentView(FrameLayout root) {
        View view = root.findViewWithTag(SLIDER_PERCENT_TAG);
        return view instanceof TextView ? (TextView) view : null;
    }

    private FrameLayout getFrameLayoutField(Object sliderView) {
        Object field = XposedHelpers.getObjectField(sliderView, BrightnessSliderPercentageHook.BRIGHTNESS_ROOT_FIELD);
        return field instanceof FrameLayout ? (FrameLayout) field : null;
    }

    private View getViewField(Object sliderView) {
        Object field = XposedHelpers.getObjectField(sliderView, BrightnessSliderPercentageHook.BRIGHTNESS_ICON_FIELD);
        return field instanceof View ? (View) field : null;
    }

    private boolean isBrightnessSlider(Object sliderView, Object view) {
        Object slider = XposedHelpers.getObjectField(sliderView, "mBrightnessSlider");
        return slider == view;
    }

    private Object findToggleSliderView(View seekBar) {
        View current = seekBar;
        for (int i = 0; i < 5 && current != null; i++) {
            ViewParent parent = current.getParent();
            if (!(parent instanceof View)) {
                break;
            }
            current = (View) parent;
            if (TOGGLE_SLIDER_VIEW_CLASS.equals(current.getClass().getName())) {
                return current;
            }
        }
        return null;
    }

    private void ensureLayoutTracking(FrameLayout root, View icon, TextView percentView) {
        if (layoutListeners.containsKey(root)) {
            root.post(() -> positionLabel(root, icon, percentView));
            return;
        }

        View.OnLayoutChangeListener listener = (v, left, top, right, bottom, oldLeft, oldTop, oldRight, oldBottom) -> positionLabel(root, icon, percentView);
        layoutListeners.put(root, listener);
        root.addOnLayoutChangeListener(listener);
        root.post(() -> positionLabel(root, icon, percentView));
    }

    private void detachBrightnessLabel(Object sliderView) {
        removePercentView(getFrameLayoutField(sliderView));
    }

    private void removePercentView(FrameLayout root) {
        if (root == null) {
            return;
        }
        View view = root.findViewWithTag(SLIDER_PERCENT_TAG);
        if (view != null) {
            root.removeView(view);
        }
        View.OnLayoutChangeListener listener = layoutListeners.remove(root);
        if (listener != null) {
            root.removeOnLayoutChangeListener(listener);
        }
    }

    private void schedulePositionUpdate(Object sliderView) {
        FrameLayout root = getFrameLayoutField(sliderView);
        View icon = getViewField(sliderView);
        if (root == null || icon == null) {
            return;
        }
        TextView percentView = findPercentView(root);
        if (percentView == null) {
            return;
        }
        root.post(() -> positionLabel(root, icon, percentView));
    }

    private void positionLabel(FrameLayout root, View icon, TextView percentView) {
        if (root == null || icon == null || percentView == null) {
            return;
        }

        int rootWidth = root.getWidth();
        int rootHeight = root.getHeight();
        if (rootWidth <= 0 || rootHeight <= 0) {
            return;
        }

        if (percentView.getMeasuredWidth() == 0 || percentView.getMeasuredHeight() == 0) {
            percentView.measure(
                    View.MeasureSpec.makeMeasureSpec(rootWidth, View.MeasureSpec.AT_MOST),
                    View.MeasureSpec.makeMeasureSpec(rootHeight, View.MeasureSpec.AT_MOST)
            );
        }

        int labelWidth = Math.max(1, percentView.getMeasuredWidth());
        int labelHeight = Math.max(1, percentView.getMeasuredHeight());
        int iconCenterX = icon.getLeft() + (icon.getWidth() / 2);
        int targetLeft = iconCenterX - (labelWidth / 2);
        int targetTop = icon.getBottom() + dp(root.getContext(), LABEL_GAP_DP);

        targetLeft = Math.max(0, Math.min(targetLeft, Math.max(0, rootWidth - labelWidth)));
        targetTop = Math.max(0, Math.min(targetTop, Math.max(0, rootHeight - labelHeight)));

        FrameLayout.LayoutParams params = (FrameLayout.LayoutParams) percentView.getLayoutParams();
        if (params == null) {
            params = createLayoutParams();
        }
        params.width = ViewGroup.LayoutParams.WRAP_CONTENT;
        params.height = ViewGroup.LayoutParams.WRAP_CONTENT;
        params.gravity = android.view.Gravity.TOP | android.view.Gravity.START;
        params.leftMargin = targetLeft;
        params.topMargin = targetTop;
        percentView.setLayoutParams(params);
    }

    private int dp(Context context, int value) {
        return Math.round(value * context.getResources().getDisplayMetrics().density);
    }

    private void updatePrefs() {
        PreferenceHelper prefs = PreferenceHelper.getInstance();
        PERCENTAGE_ENABLED = prefs.getBoolean(PREF_KEY, false);
    }
}

File: app/src/main/java/com/qimian233/ztool/hook/modules/systemui/qs/BrightnessSliderPercentageHook.kt
package com.qimian233.ztool.hook.modules.systemui.qs

import android.annotation.SuppressLint
import android.content.Context
import android.graphics.Color
import android.graphics.Typeface
import android.text.TextUtils
import android.util.AttributeSet
import android.view.Gravity
import android.view.View
import android.view.ViewGroup
import android.widget.FrameLayout
import android.widget.ProgressBar
import android.widget.SeekBar
import android.widget.TextView
import com.qimian233.ztool.data.keys.PreferenceKeys
import com.qimian233.ztool.data.keys.ScopeKeys
import com.qimian233.ztool.hook.base.AppHookModule
import io.github.libxposed.api.XposedModuleInterface.PackageLoadedParam
import java.lang.reflect.Constructor
import java.lang.reflect.Method
import java.util.Locale
import java.util.WeakHashMap
import kotlin.math.roundToInt

@SuppressLint("PrivateApi")
class BrightnessSliderPercentageHook : AppHookModule() {

    private val layoutListeners = WeakHashMap<View, View.OnLayoutChangeListener>()
    private val pendingPositionUpdates = WeakHashMap<FrameLayout, Boolean>()

    override fun getModuleName(): String = PreferenceKeys.BRIGHTNESS_SLIDER_PERCENTAGE.name

    override fun getTargetPackages(): Array<String> = arrayOf(SYSTEM_UI_PACKAGE)

    override fun handleLoadPackage(param: PackageLoadedParam) {
        val classLoader = param.defaultClassLoader
        updatePrefs()
        hookToggleSliderViewLifecycle(classLoader)
        hookBrightnessControllerCallbacks(classLoader)
        hookSeekProgressChanges()
        logger.info("Brightness slider percentage hooks installed")
    }

    private fun hookBrightnessControllerCallbacks(classLoader: ClassLoader) {
        try {
            val onChangedMethod: Method = classLoader
                .loadClass("com.android.systemui.settings.brightness.BrightnessController")
                .getDeclaredMethod(
                    "onChanged",
                    Int::class.javaPrimitiveType,
                    Boolean::class.javaPrimitiveType,
                    Boolean::class.javaPrimitiveType
                )
            hookWithId(onChangedMethod, "on_changed") { chain ->
                val result = chain.proceed()
                refreshBrightnessFromController(chain.thisObject, chain.args[0] as Int)
                result
            }
        } catch (_: Throwable) {
        }

        try {
            val setValueMethod: Method = classLoader
                .loadClass("com.android.systemui.settings.brightness.BrightnessSliderController")
                .getDeclaredMethod("setValue", Int::class.javaPrimitiveType)
            hookWithId(setValueMethod, "set_value") { chain ->
                val result = chain.proceed()
                val sliderController = chain.thisObject
                val scCls = sliderController.javaClass
                try {
                    scCls.getDeclaredField("mBrightnessSliderHapticPlugin")
                } catch (_: NoSuchFieldException) {
                    logger.warn("Field mBrightnessSliderHapticPlugin not found, shouldn't hook mView of this class!")
                    return@hookWithId chain.proceed()
                }
                val view = findField(scCls, "mView").get(sliderController)
                if (view is View) {
                    refreshBrightnessFromView(view, chain.args[0] as Int)
                }
                result
            }
        } catch (_: Throwable) {
        }
    }

    private fun hookToggleSliderViewLifecycle(classLoader: ClassLoader) {
        try {
            val ctor: Constructor<*> = classLoader.loadClass(TOGGLE_SLIDER_VIEW_CLASS)
                .getDeclaredConstructor(Context::class.java, AttributeSet::class.java, Int::class.javaPrimitiveType)
            hookWithId(ctor, "ctor") { chain ->
                chain.proceed()
                attachSliderLabel(chain.thisObject)
                null
            }
        } catch (_: Throwable) {
        }

        try {
            val updateBrightnessMethod: Method = classLoader.loadClass(TOGGLE_SLIDER_VIEW_CLASS)
                .getDeclaredMethod("updateBrightnessSlider")
            hookWithId(updateBrightnessMethod, "update_brightness") { chain ->
                val result = chain.proceed()
                attachSliderLabel(chain.thisObject)
                refreshBrightnessLabel(chain.thisObject)
                result
            }
        } catch (_: Throwable) {
        }

        try {
            val refreshSeekBarMethod: Method = classLoader.loadClass(TOGGLE_SLIDER_VIEW_CLASS)
                .getDeclaredMethod("refreshSeekBar", ProgressBar::class.java)
            hookWithId(refreshSeekBarMethod, "refresh_seek_bar") { chain ->
                val result = chain.proceed()
                val sliderView = chain.thisObject
                val progressBar = chain.args[0] as ProgressBar
                if (isBrightnessSlider(sliderView, progressBar)) {
                    refreshBrightnessLabel(sliderView)
                }
                result
            }
        } catch (_: Throwable) {
        }
    }

    private fun hookSeekProgressChanges() {
        try {
            val setProgressMethod: Method =
                SeekBar::class.java.getDeclaredMethod("setProgress", Int::class.javaPrimitiveType)
            hookWithId(setProgressMethod, "set_progress") { chain ->
                val result = chain.proceed()
                val seekBar = chain.thisObject as SeekBar
                val sliderView = findToggleSliderView(seekBar) ?: return@hookWithId result
                if (isBrightnessSlider(sliderView, seekBar)) {
                    refreshBrightnessLabel(sliderView)
                }
                result
            }
        } catch (_: Throwable) {
        }
    }

    private fun attachSliderLabel(sliderView: Any) {
        try {
            attachLabelToRoot(sliderView)
        } catch (t: Throwable) {
            logger.error("Failed to attach brightness slider label", t)
        }
    }

    private fun attachLabelToRoot(sliderView: Any) {
        val root = getFrameLayoutField(sliderView)
        val icon = getViewField(sliderView)
        if (root == null || icon == null) {
            return
        }

        if (!percentageEnabled) {
            removePercentView(root)
            return
        }

        var percentView = findPercentView(root)
        if (percentView == null) {
            percentView = createPercentView(root.context)
            root.addView(percentView, createLayoutParams())
        }

        ensureLayoutTracking(root, icon, percentView)
        updateBrightnessPercentColor(sliderView, percentView)
        refreshBrightnessLabel(sliderView, percentView)
    }

    private fun refreshBrightnessLabel(sliderView: Any) {
        if (!percentageEnabled) {
            detachBrightnessLabel(sliderView)
            return
        }
        val root = getFrameLayoutField(sliderView) ?: return
        val percentView = findPercentView(root)
        if (percentView != null) {
            refreshBrightnessLabel(sliderView, percentView)
        }
    }

    private fun refreshBrightnessLabel(sliderView: Any, percentView: TextView) {
        if (!percentageEnabled) {
            removePercentView(getFrameLayoutField(sliderView))
            return
        }
        try {
            val root = getFrameLayoutField(sliderView)
            if (root != null && root.isInLayout) {
                schedulePositionUpdate(sliderView)
                return
            }
            val cl = sliderView.javaClass
            val brightnessSlider = cl.getDeclaredField("mBrightnessSlider").get(sliderView) as SeekBar
            setPercentTextIfChanged(
                percentView,
                formatPercent(
                    brightnessSlider.progress,
                    brightnessSlider.min,
                    brightnessSlider.max
                )
            )
            updateBrightnessPercentColor(sliderView, percentView)
            schedulePositionUpdate(sliderView)
        } catch (t: Throwable) {
            setPercentTextIfChanged(percentView, "--%")
            logger.error("Failed to refresh brightness percent", t)
        }
    }

    private fun refreshBrightnessFromController(brightnessController: Any, progress: Int) {
        if (!percentageEnabled) {
            return
        }
        try {
            val bcCls = brightnessController.javaClass
            val control = bcCls.getDeclaredField("mControl").get(brightnessController) ?: return
            try {
                bcCls.getDeclaredField("mBrightnessObserver")
            } catch (_: NoSuchFieldException) {
                logger.warn("Field mBrightnessObserver not found, shouldn't hook mView of this class!")
                return
            }
            val view = findField(control.javaClass, "mView").get(control)
            if (view !is View) {
                return
            }
            refreshBrightnessFromView(view, progress)
        } catch (t: Throwable) {
            logger.error("Failed to refresh brightness from controller", t)
        }
    }

    private fun refreshBrightnessFromView(view: View, progress: Int) {
        val sliderView = findToggleSliderView(view) ?: return
        val root = getFrameLayoutField(sliderView) ?: return
        val percentView = findPercentView(root) ?: return

        try {
            val cl = sliderView.javaClass
            val brightnessSlider = cl.getDeclaredField("mBrightnessSlider").get(sliderView) as SeekBar
            var max = brightnessSlider.max
            max = 1.coerceAtLeast(max)
            val percent = 0.coerceAtLeast(100.coerceAtMost(((progress * 100f) / max).roundToInt()))
            if (root.isInLayout) {
                schedulePositionUpdate(sliderView)
                return
            }
            setPercentTextIfChanged(percentView, String.format(Locale.US, "%d%%", percent))
            updateBrightnessPercentColor(sliderView, percentView)
            schedulePositionUpdate(sliderView)
        } catch (t: Throwable) {
            logger.error("Failed to refresh brightness from view", t)
        }
    }

    private fun updateBrightnessPercentColor(sliderView: Any, percentView: TextView) {
        percentView.setTextColor(resolveBrightnessPercentColor(sliderView))
    }

    private fun resolveBrightnessPercentColor(sliderView: Any): Int {
        val brightnessSlider: SeekBar?
        try {
            brightnessSlider = sliderView as? SeekBar
                ?: sliderView.javaClass.getDeclaredField("mBrightnessSlider").get(sliderView) as SeekBar
        } catch (_: Throwable) {
            return Color.argb(0xff, 0xd8, 0xd8, 0xd8)
        }
        val progress = ((brightnessSlider.progress - brightnessSlider.min) * 1.0f) /
                1.coerceAtLeast(brightnessSlider.max - brightnessSlider.min)
        if (progress < 0.2f) {
            return Color.argb(0xff, 0xd8, 0xd8, 0xd8)
        }
        var gray = ((1.0f - ((progress - 0.2f) / 0.2f).coerceAtMost(1.0f)) * 216.0f).toInt()
        gray = gray.coerceAtLeast(0x80)
        val alpha = (kotlin.math.floor(progress * 85.0f).toInt() + 170).coerceAtMost(255)
        return Color.argb(alpha, gray, gray, gray)
    }

    private fun formatPercent(progress: Int, min: Int, max: Int): String {
        val range = 1.coerceAtLeast(max - min)
        var value = (((progress - min) * 100f) / range).roundToInt()
        value = 0.coerceAtLeast(100.coerceAtMost(value))
        return String.format(Locale.US, "%d%%", value)
    }

    private fun createPercentView(context: Context): TextView {
        val textView = TextView(context)
        textView.tag = SLIDER_PERCENT_TAG
        textView.setTextColor(Color.argb(0xff, 0xd8, 0xd8, 0xd8))
        textView.setTypeface(Typeface.DEFAULT_BOLD)
        textView.textSize = 13f
        textView.setShadowLayer(2f, 0f, 0f, Color.BLACK)
        textView.isSingleLine = true
        textView.includeFontPadding = false
        textView.gravity = Gravity.CENTER
        textView.importantForAccessibility = View.IMPORTANT_FOR_ACCESSIBILITY_NO
        textView.isClickable = false
        textView.isFocusable = false
        textView.setPadding(dp(context, 6), dp(context, 2), dp(context, 6), dp(context, 2))
        textView.elevation = dp(context, 2).toFloat()
        return textView
    }

    private fun createLayoutParams(): FrameLayout.LayoutParams {
        val params = FrameLayout.LayoutParams(
            ViewGroup.LayoutParams.WRAP_CONTENT,
            ViewGroup.LayoutParams.WRAP_CONTENT
        )
        params.gravity = Gravity.TOP or Gravity.START
        return params
    }

    private fun findPercentView(root: FrameLayout): TextView? {
        val view = root.findViewWithTag<View>(SLIDER_PERCENT_TAG)
        return view as? TextView
    }

    private fun getFrameLayoutField(sliderView: Any): FrameLayout? {
        return try {
            val field = sliderView.javaClass.getDeclaredField(BRIGHTNESS_ROOT_FIELD).get(sliderView)
            field as? FrameLayout
        } catch (_: Throwable) {
            null
        }
    }

    private fun getViewField(sliderView: Any): View? {
        return try {
            val field = sliderView.javaClass.getDeclaredField(BRIGHTNESS_ICON_FIELD).get(sliderView)
            field as? View
        } catch (_: Throwable) {
            null
        }
    }

    private fun isBrightnessSlider(sliderView: Any, view: Any): Boolean {
        return try {
            val slider = sliderView.javaClass.getDeclaredField("mBrightnessSlider").get(sliderView)
            slider == view
        } catch (_: Throwable) {
            false
        }
    }

    private fun findToggleSliderView(seekBar: View): Any? {
        var current: View = seekBar
        for (i in 0 until 5) {
            val parent = current.parent
            if (parent !is View) {
                break
            }
            current = parent
            if (TOGGLE_SLIDER_VIEW_CLASS == current.javaClass.name) {
                return current
            }
        }
        return null
    }

    private fun ensureLayoutTracking(root: FrameLayout, icon: View, percentView: TextView) {
        if (layoutListeners.containsKey(root)) {
            schedulePositionUpdate(root, icon, percentView)
            return
        }

        val listener = View.OnLayoutChangeListener { _, _, _, _, _, _, _, _, _ ->
            schedulePositionUpdate(root, icon, percentView)
        }
        layoutListeners[root] = listener
        root.addOnLayoutChangeListener(listener)
        schedulePositionUpdate(root, icon, percentView)
    }

    private fun detachBrightnessLabel(sliderView: Any) {
        removePercentView(getFrameLayoutField(sliderView))
    }

    private fun removePercentView(root: FrameLayout?) {
        if (root == null) {
            return
        }
        val view = root.findViewWithTag<View>(SLIDER_PERCENT_TAG)
        if (view != null) {
            root.removeView(view)
        }
        val listener = layoutListeners.remove(root)
        if (listener != null) {
            root.removeOnLayoutChangeListener(listener)
        }
    }

    private fun schedulePositionUpdate(sliderView: Any) {
        val root = getFrameLayoutField(sliderView)
        val icon = getViewField(sliderView)
        if (root == null || icon == null) {
            return
        }
        val percentView = findPercentView(root) ?: return
        schedulePositionUpdate(root, icon, percentView)
    }

    private fun schedulePositionUpdate(root: FrameLayout, icon: View, percentView: TextView) {
        if (java.lang.Boolean.TRUE == pendingPositionUpdates[root]) {
            return
        }
        pendingPositionUpdates[root] = java.lang.Boolean.TRUE
        root.post {
            try {
                positionLabel(root, icon, percentView)
            } finally {
                pendingPositionUpdates.remove(root)
            }
        }
    }

    private fun positionLabel(root: FrameLayout, icon: View, percentView: TextView) {

        val rootWidth = root.width
        val rootHeight = root.height
        if (rootWidth <= 0 || rootHeight <= 0) {
            return
        }

        if (percentView.measuredWidth == 0 || percentView.measuredHeight == 0) {
            percentView.measure(
                View.MeasureSpec.makeMeasureSpec(rootWidth, View.MeasureSpec.AT_MOST),
                View.MeasureSpec.makeMeasureSpec(rootHeight, View.MeasureSpec.AT_MOST)
            )
        }

        val labelWidth = 1.coerceAtLeast(percentView.measuredWidth)
        val labelHeight = 1.coerceAtLeast(percentView.measuredHeight)
        val iconCenterX = icon.left + (icon.width / 2)
        var targetLeft = iconCenterX - (labelWidth / 2)
        var targetTop = icon.bottom + dp(root.context, LABEL_GAP_DP)

        targetLeft = 0.coerceAtLeast(targetLeft.coerceAtMost(0.coerceAtLeast(rootWidth - labelWidth)))
        targetTop = 0.coerceAtLeast(targetTop.coerceAtMost(0.coerceAtLeast(rootHeight - labelHeight)))

        var params = percentView.layoutParams as? FrameLayout.LayoutParams
        if (params == null) {
            params = createLayoutParams()
        }
        if (params.leftMargin == targetLeft && params.topMargin == targetTop
            && params.width == ViewGroup.LayoutParams.WRAP_CONTENT
            && params.height == ViewGroup.LayoutParams.WRAP_CONTENT
            && params.gravity == (Gravity.TOP or Gravity.START)
        ) {
            return
        }
        params.width = ViewGroup.LayoutParams.WRAP_CONTENT
        params.height = ViewGroup.LayoutParams.WRAP_CONTENT
        params.gravity = Gravity.TOP or Gravity.START
        params.leftMargin = targetLeft
        params.topMargin = targetTop
        percentView.layoutParams = params
    }

    private fun setPercentTextIfChanged(percentView: TextView, text: String) {
        if (TextUtils.equals(percentView.text, text)) {
            return
        }
        percentView.text = text
    }

    private fun dp(context: Context, value: Int): Int {
        return (value * context.resources.displayMetrics.density).roundToInt()
    }

    private fun updatePrefs() {
        percentageEnabled = try {
            remotePreferences.getBoolean(PreferenceKeys.BRIGHTNESS_SLIDER_PERCENTAGE.name, false)
        } catch (_: Throwable) {
            false
        }
    }

    private var percentageEnabled = false

    companion object {
        private val SYSTEM_UI_PACKAGE = ScopeKeys.SYSTEM_UI.packageName
        private const val TOGGLE_SLIDER_VIEW_CLASS = "com.android.systemui.settings.ToggleSliderView"
        private const val SLIDER_PERCENT_TAG = "ztool_control_center_slider_percent"
        private const val BRIGHTNESS_ROOT_FIELD = "mBrightnessSliderRoot"
        private const val BRIGHTNESS_ICON_FIELD = "mBrightnessIconMark"
        private const val LABEL_GAP_DP = 2
    }
}
